.github/workflows/ci.yml

```diff
@@ -39,4 +37,67 @@ jobs:
+            cd examples/KcMwm && echo "KcMwm CPP Demo:"  
+            cmake -B build && cmake --build build && cd build && ./demo
+            cd .. && cd Tsp && echo "Tsp CPP Demo:"
+            cmake -B build && cmake --build build && cd build && ./demo
+            cd .. && cd Ksat && echo "Ksat CPP Demo:" 
+            cmake -B build && cmake --build build && cd build && ./demo
+            echo "CPP Demos executed successfully."
+    - name: Run Python Demos
+      run: |
+            python3 -m pip install --user . -v
+            cd src/impa
+            echo "Python (CPP) Wrapper Demos:"
+            python3 main_kc_mwm.py --nITER=400 --filteringFlag=True --alpha=0.9 --PPFlag=True --PPOption=1 --threshold=-0.0001
+            python3 main_tsp.py --filteringFlag=True --alpha=0.5 --augmFlag=True --threshold=-0.0001 --nITER=200 --inputPath=inputs_1000_nNodes_random --testFile=0 --lkhSolFlag=True --maxAugmCount=10
+            python3 main_tsp.py --nNodes=10 --filteringFlag=True --alpha=0.5 --augmFlag=True --threshold=-0.0001 --nITER=200 --randomTestFlag=True --lkhSolFlag=True --maxAugmCount=20
+            python3 main_ksat.py --filteringFlag=True --threshold=-0.0001 --nITER=200 --inputPath=inputs_ksat_zeromean_var_10_samples_500 --outputPath=outputs_ksat_zeromean_var_10_samples_500 --PPElements=2 --testFile=112 --alpha=0.5 --var=8 --typeMetrics=2 --overwrite=1 --PPFlag=True
+            python3 main_ksat.py --filteringFlag=True --threshold=-0.0001 --nITER=200 --PPElements=2 --alpha=0.5 --var=8 --randomTestFlag=True --nConstraints=200 --nVariables=90 --kVariable=5 --PPFlag=True
+            cd ../.. && cd test/python_ksat/src && echo "Ksat Python Demo:"
+            python3 main_ksat.py --filteringFlag=True --threshold=-0.0001 --nITER=200 --alpha=0.5 --randomTestFlag=True --nConstraints=20 --nVariables=10 --kVariable=3
+            cd ../.. && cd python_tsp/src && echo "Tsp Python Demo:"
+            python3 main_tsp.py --nNodes=10 --filteringFlag=True --alpha=0.5 --augmFlag=True --threshold=-0.0001 --nITER=200 --randomTestFlag=True
+            cd ../.. && cd python_kc_mwm/src && echo "KcMwm Python Demo:"
+            python3 main_pure_optimized.py --nITER=400 --filteringFlag=True --alpha=0.9 --PPFlag=True --threshold=-0.0001
+            echo "Python Demos executed successfully."
+
+  test:
+    name: Unit Test Job
+    runs-on: ${{ matrix.os }}
+    strategy: 
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+    steps:
+    - uses: actions/checkout@v4
+    - uses: actions/setup-python@v5
+      with: 
+        python-version: '3.9'
+        cache: 'pip'
+        cache-dependency-path: setup.py
+    - name: Run Unit Tests
+      run: |
+            python3 -m pip install --user . -v
+            cmake -B build && cmake --build build && cd build/test/src
+            ./unit_test_kc_mwm.sh
+            ./unit_test_tsp.sh
+            ./unit_test_ksat.sh
+            echo "Unit tests completed successfully."
```

Comment:
The goal of these checks is not only to have unit tests completed successfully but also to ensure that they pass. Can we verify this in this workflow without manual verification?

Reply:
I should be able to add a check for is the scripts fail. Looking into this

Reply:
@RustomAlexios 

I've gone ahead and adjusted the `ut_utils.py` scripts to write to a "log" `.txt` file of sorts (for each `ksat`, `tsp` and `kcmwm` jobs) and updated the `ci.yml` file to `grep` for the keyword "FAILED". 

If any one of the sub unit tests have failed this will cause the job to fail. 

Let me know what you think